components/tidb_query/src/codec/datum.rs

```diff
@@ -784,40 +787,41 @@ impl From<Json> for Datum {
     }
 }
 
-/// `decode_datum` decodes on a datum from a byte slice generated by tidb.
-pub fn decode_datum(data: &mut BytesSlice<'_>) -> Result<Datum> {
-    if !data.is_empty() {
-        let flag = data[0];
-        *data = &data[1..];
+/// `DatumDecoder` decodes the datum.
+pub trait DatumDecoder:
+    DecimalDecoder + JsonDecoder + CompactByteDecoder + MemComparableByteDecoder
+{
+    /// `decode_datum` decodes on a datum from a byte slice generated by TiDB.
+    fn decode_datum(&mut self) -> Result<Datum> {
+        let flag = self.read_u8()?;
         let datum = match flag {
-            INT_FLAG => number::decode_i64(data).map(Datum::I64)?,
-            UINT_FLAG => number::decode_u64(data).map(Datum::U64)?,
-            BYTES_FLAG => bytes::decode_bytes(data, false).map(Datum::Bytes)?,
-            COMPACT_BYTES_FLAG => bytes::decode_compact_bytes(data).map(Datum::Bytes)?,
+            INT_FLAG => self.read_i64().map(Datum::I64)?,
+            UINT_FLAG => self.read_u64().map(Datum::U64)?,
+            BYTES_FLAG => self.read_comparable_bytes().map(Datum::Bytes)?,
+            COMPACT_BYTES_FLAG => self.read_compact_bytes().map(Datum::Bytes)?,
             NIL_FLAG => Datum::Null,
-            FLOAT_FLAG => number::decode_f64(data).map(Datum::F64)?,
+            FLOAT_FLAG => self.read_f64().map(Datum::F64)?,
             DURATION_FLAG => {
-                let nanos = number::decode_i64(data)?;
+                let nanos = self.read_i64()?;
                 let dur = Duration::from_nanos(nanos, MAX_FSP)?;
                 Datum::Dur(dur)
             }
-            DECIMAL_FLAG => data.decode_decimal().map(Datum::Dec)?,
-            VAR_INT_FLAG => number::decode_var_i64(data).map(Datum::I64)?,
-            VAR_UINT_FLAG => number::decode_var_u64(data).map(Datum::U64)?,
-            JSON_FLAG => data.decode_json().map(Datum::Json)?,
+            DECIMAL_FLAG => self.decode_decimal().map(Datum::Dec)?,
```

Comment:
Also rename the encode parts to `write_xxx`?  `write_json`/`read_json` etc.

Reply:
@niedhui Yes, I think so. Note that `read_` / `write_` prefix was only used for distinguish the new codec and the old codec (since their signatures are the same and can cause name conflicts).

Reply:
Got it. I was wondering why the method name of `NumberDecoder`  starts with `read_` before.
Code pushed, PTAL